androlo/bytes/ByteSlice.sol

```diff
@@ -0,0 +1,171 @@
+/// Used to create slices from 'bytes'.
+contract ByteSlice {
+
+    struct Slice {
+        uint _unsafe_memPtr;   // Memory address of the first byte.
+        uint _unsafe_len;      // Length.
+    }
+
+    /// Create a slice from bytes.
+    function fromBytes(bytes memory bts) internal constant returns (Slice memory slice) {
+        uint memPtr;
+        assembly {
+            memPtr := add(bts, 0x20)
+        }
+        slice = Slice(memPtr, bts.length);
+    }
+
+    /// Length of the slice.
+    function len(Slice memory slice) internal constant returns (uint) {
+        return slice._unsafe_len;
+    }
+
+    /// returns the byte at 'index'.
+    /// Index must be less then 'len(slice)', or the function will throw.
+    function at(Slice memory slice, uint index) internal constant returns (byte b) {
+        if (index >= slice._unsafe_len)
+            throw;
+        uint start = slice._unsafe_memPtr;
+        uint bb;
+        assembly {
+            bb := byte(0, mload(add(start, index)))
+        }
+        b = byte(bb);
+    }
+
+    /// returns the byte at index 'index', where 'index' is allowed to be negative.
+    /// The index must be less then 'len(slice)', and greater then '-len(slice)'
+    function at(Slice memory slice, int index) internal constant returns (byte b) {
+        if (index >= 0)
+            return at(slice, uint(index));
+        uint iAbs = uint(-index);
+        if (iAbs > slice._unsafe_len)
+            throw;
+        return at(slice, slice._unsafe_len - iAbs);
+    }
+
+    /// creates a 'bytes' memory-variable from a slice, copying the data.
+    function toBytes(Slice memory slice) internal constant returns (bytes memory bts) {
+        uint len = slice._unsafe_len;
+        uint memPtr = slice._unsafe_memPtr;
+        bts = new bytes(len);
+        if (len == 0)
+            return;
+        // We can do word-by-word copying since 'bts' was the last thing to be
+        // allocated. Just overwrite any excess bytes with zeroes at the end.
+        assembly {
+            {
+                    let i := 0
+                    let btsOffset := add(bts, 0x20)
+                    let words := add(add(div(len, 32), gt(mod(len, 32), 0)), 1)
+                tag_loop:
+                    jumpi(end, gt(i, words))
+                    {
+                        let offset := mul(i, 32)
+                        mstore(add(btsOffset, offset), mload(add(memPtr, offset)))
+                        i := add(i, 1)
+                    }
+                    jump(tag_loop)
+                end:
+                    mstore(add(add(bts, len), 0x20), 0)
+            }
+        }
+    }
+
+    /// Copy the slice.
+    function newSlice(Slice memory slice) internal constant returns (Slice memory newSlice) {
```

Comment:
Actually, I think what it means is perhaps to avoid using `Slice()` altogether since the slice is already created. Cost of `newSlice(Slice memory slice)` went down from 136 to 106 when setting each field individually.

@chriseth am i right?